avr-hal-generic/src/spi.rs

```diff
@@ -64,220 +66,258 @@ impl Default for Settings {
+/// ownership of the MOSI and MISO pins to ensure they are in the correct mode.
+/// Instantiate with the `new` method.
+pub struct Spi<H, SPI, SCLK, MOSI, MISO, CS> {
+    p: SPI,
+    sclk: SCLK,
+    mosi: MOSI,
+    miso: MISO,
+    _cs: PhantomData<CS>,
+    _h: PhantomData<H>,
+}
 
-        impl $Spi<$crate::port::mode::Floating> {
-            /// Instantiate an SPI with the registers, SCLK/MOSI/MISO/CS pins, and settings,
-            /// with an external pull-up on the MISO pin.
-            ///
-            /// The pins are not actually used directly, but they are moved into the struct in
-            /// order to enforce that they are in the correct mode, and cannot be used by anyone
-            /// else while SPI is active.  CS is placed into a `ChipSelectPin` instance and given
-            /// back so that its output state can be changed as needed.
-            pub fn with_external_pullup(
-                peripheral: $SPI,
-                sclk: $sclkmod::$SCLK<$crate::port::mode::Output>,
-                mosi: $mosimod::$MOSI<$crate::port::mode::Output>,
-                miso: $misomod::$MISO<$crate::port::mode::Input<$crate::port::mode::Floating>>,
-                settings: Settings
-            ) -> Self {
-                let spi = $Spi {
-                    peripheral,
-                    sclk,
-                    mosi,
-                    miso,
-                    settings,
-                    is_write_in_progress: false,
-                };
-                spi.setup();
-                spi
-            }
-        }
+impl<H, SPI, SCLKPIN, MOSIPIN, MISOPIN, CSPIN>
+    Spi<
+        H,
+        SPI,
+        port::Pin<port::mode::Output, SCLKPIN>,
+        port::Pin<port::mode::Output, MOSIPIN>,
+        port::Pin<port::mode::Input, MISOPIN>,
+        port::Pin<port::mode::Output, CSPIN>,
+    >
+where
+    SPI: SpiOps<
+        H,
+        port::Pin<port::mode::Output, SCLKPIN>,
+        port::Pin<port::mode::Output, MOSIPIN>,
+        port::Pin<port::mode::Input, MISOPIN>,
+        port::Pin<port::mode::Output, CSPIN>,
+    >,
+    SCLKPIN: port::PinOps,
+    MOSIPIN: port::PinOps,
+    MISOPIN: port::PinOps,
+    CSPIN: hal::digital::v2::OutputPin,
+{
+    /// Instantiate an SPI with the registers, SCLK/MOSI/MISO/CS pins, and settings,
+    /// with the internal pull-up enabled on the MISO pin.
+    ///
+    /// The pins are not actually used directly, but they are moved into the struct in
+    /// order to enforce that they are in the correct mode, and cannot be used by anyone
+    /// else while SPI is active.  CS is placed into a `ChipSelectPin` instance and given
+    /// back so that its output state can be changed as needed.
+    pub fn new(
+        p: SPI,
+        sclk: port::Pin<port::mode::Output, SCLKPIN>,
+        mosi: port::Pin<port::mode::Output, MOSIPIN>,
+        miso: port::Pin<port::mode::Input<port::mode::PullUp>, MISOPIN>,
+        cs: CSPIN,
```

Comment:
I think this should also be `port::Pin<port::mode::Output, CSPIN>` here and everywhere else?

Reply:
I think I fixed this; it compiles, at least?